src/inc/til/rle.h

```diff
@@ -0,0 +1,1064 @@
+                end->length -= end_pos;
+            }
+
+            // NOTE: It's possible for begin > end, as we increment begin in [Step3].
+            const size_t available_space = begin < end ? end - begin : 0;
+            const size_t required_space = replacements.size() + (mid_insertion_trailer ? 1 : 0);
+            const auto begin_index = begin - _runs.begin();
+            const auto replacements_begin = replacements.begin();
+            const auto replacements_end = replacements.end();
+
+            // [Step4]
+            const auto direct_copy_end = replacements_begin + std::min(available_space, replacements.size());
+            begin = std::copy(replacements_begin, direct_copy_end, begin);
+
+            if (available_space >= required_space)
+            {
+                // [Step6.1]
+                _runs.erase(begin, end);
+            }
+            else
+            {
+                if (mid_insertion_trailer)
+                {
+                    // Unfortunately there's no efficient way to express "insert an iterator range
+                    // plus one extra element at the end" with standard vector containers.
+                    // --> First make some space for N+1 elements using default initialization.
+                    //     Then insert the new runs and finally the mid_insertion_trailer.
+                    _runs.insert(begin, required_space - available_space, {});
+                    // [Step6.2]
+                    begin = std::copy(direct_copy_end, replacements_end, _runs.begin() + begin_index);
+                    // [Step5]
+                    *begin = *std::move(mid_insertion_trailer);
+                }
+                else
+                {
+                    // [Step6.2]
+                    _runs.insert(begin, direct_copy_end, replacements_end);
+                }
+            }
+
+            // [Step7]
+            if (begin_additional_length)
+            {
+                begin = _runs.begin() + begin_index;
+                begin->length += begin_additional_length;
+            }
+            if (end_additional_length)
+            {
+                end = _runs.begin() + begin_index + required_space - 1;
+                end->length += end_additional_length;
+            }
+
+            // [Step8]
+            _total_length -= end_index - start_index;
+            for (const auto& run : replacements)
+            {
+                _total_length += run.length;
+            }
+        }
+
+        container _runs;
+        S _total_length{ 0 };
+
+#ifdef UNIT_TESTING
+        friend class ::RunLengthEncodingTests;
+#endif
+    };
+
+    template<typename T, typename S = std::size_t>
+    using rle = basic_rle<T, S, std::vector<rle_pair<T, S>>>;
+
+#ifdef BOOST_CONTAINER_CONTAINER_SMALL_VECTOR_HPP
+    template<typename T, typename S = std::size_t, std::size_t N = 1>
+    using small_rle = basic_rle<T, S, boost::container::small_vector<rle_pair<T, S>, N>>;
+#endif
+};

```

Comment:
If you wanted to add one more class template (for fun!) i would suggest `til::pmr::rle` that uses a `std::pmr::vector`. However! basic_rle can't take an Allocator type, and this seems like an unnecessary cost for something we do not currently need. 😁

Reply:
Ah, I meant "type alias" instead of "class template"